File: frontend/js/volunteers.js
let allVolunteers = [];
let editingVolunteerId = null;


/* =========================
   PAGE LOAD
========================= */

document.addEventListener("DOMContentLoaded", () => {
    loadVolunteers();
});


/* =========================
   LOAD VOLUNTEERS
========================= */

async function loadVolunteers() {

    hideError();

    const tableBody = document.getElementById("volunteers-table-body");

    tableBody.innerHTML = `
        <tr>
            <td colspan="8" class="loading">
                Loading volunteers...
            </td>
        </tr>
    `;

    try {

        const response = await fetch(`${API_BASE_URL}/api/volunteers/`);

        const data = await response.json();

        if (!response.ok) {
            throw new Error(
                data.detail || "Failed to load volunteers"
            );
        }

        allVolunteers = Array.isArray(data.volunteers)
            ? data.volunteers
            : [];

        updateStatistics();
        applyFilters();

    } catch (error) {

        console.error("Error loading volunteers:", error);

        tableBody.innerHTML = `
            <tr>
                <td colspan="8" class="empty-state">
                    Failed to load volunteers.
                </td>
            </tr>
        `;

        showError(error.message);

    }
}


/* =========================
   STATISTICS
========================= */

function updateStatistics() {

    const total = allVolunteers.length;

    const available = allVolunteers.filter(
        volunteer =>
            volunteer.availability_status === "AVAILABLE"
    ).length;

    const assigned = allVolunteers.filter(
        volunteer =>
            volunteer.availability_status === "ASSIGNED"
    ).length;

    const unavailable = allVolunteers.filter(
        volunteer =>
            volunteer.availability_status === "UNAVAILABLE"
    ).length;


    setText("total-count", total);
    setText("available-count", available);
    setText("assigned-count", assigned);
    setText("unavailable-count", unavailable);
}


/* =========================
   SEARCH + FILTER
========================= */

function applyFilters() {

    const searchInput =
        document.getElementById("search-input");

    const statusFilter =
        document.getElementById("status-filter");

    const searchTerm =
        searchInput.value.trim().toLowerCase();

    const selectedStatus =
        statusFilter.value;


    const filteredVolunteers = allVolunteers.filter(
        volunteer => {

            const matchesSearch =
                String(volunteer.volunteer_id ?? "")
                    .toLowerCase()
                    .includes(searchTerm) ||

                String(volunteer.user_id ?? "")
                    .toLowerCase()
                    .includes(searchTerm) ||

                String(volunteer.full_name ?? "")
                    .toLowerCase()
                    .includes(searchTerm) ||

                String(volunteer.email ?? "")
                    .toLowerCase()
                    .includes(searchTerm) ||

                String(volunteer.skills ?? "")
                    .toLowerCase()
                    .includes(searchTerm) ||

                String(volunteer.phone ?? "")
                    .toLowerCase()
                    .includes(searchTerm);


            const matchesStatus =
                selectedStatus === "ALL" ||
                volunteer.availability_status === selectedStatus;


            return matchesSearch && matchesStatus;
        }
    );


    renderVolunteers(filteredVolunteers);
}


/* =========================
   RENDER TABLE
========================= */

function renderVolunteers(volunteers) {

    const tableBody =
        document.getElementById("volunteers-table-body");


    if (volunteers.length === 0) {

        tableBody.innerHTML = `
            <tr>
                <td colspan="8" class="empty-state">
                    No volunteers found.
                </td>
            </tr>
        `;

        return;
    }


    tableBody.innerHTML = volunteers.map(
        volunteer => {

            const status =
                volunteer.availability_status || "AVAILABLE";

            const statusClass =
                getStatusClass(status);

            const skills =
                volunteer.skills
                    ? escapeHTML(volunteer.skills)
                    : "—";

            const emergencyContact =
                volunteer.emergency_contact
                    ? escapeHTML(volunteer.emergency_contact)
                    : "—";

            const fullName =
                volunteer.full_name
                    ? escapeHTML(volunteer.full_name)
                    : "Unknown User";

            const email =
                volunteer.email
                    ? escapeHTML(volunteer.email)
                    : "—";

            const phone =
                volunteer.phone
                    ? escapeHTML(volunteer.phone)
                    : "—";

            const joinedDate =
                formatDate(volunteer.joined_at);


            return `
                <tr>

                    <td>
                        ${escapeHTML(volunteer.volunteer_id)}
                    </td>


                    <td>

                        <div class="volunteer-name">

                            <strong>
                                ${fullName}
                            </strong>

                            <small>
                                User ID: ${escapeHTML(volunteer.user_id)}
                            </small>

                        </div>

                    </td>


                    <td>

                        <div class="contact-info">

                            <span>
                                ${email}
                            </span>

                            <small>
                                ${phone}
                            </small>

                        </div>

                    </td>


                    <td>

                        <div class="skills-text">
                            ${skills}
                        </div>

                    </td>


                    <td>

                        <span class="badge ${statusClass}">
                            ${escapeHTML(status)}
                        </span>

                    </td>


                    <td>

                        <span class="emergency-contact">
                            ${emergencyContact}
                        </span>

                    </td>


                    <td>

                        <span class="joined-date">
                            ${joinedDate}
                        </span>

                    </td>


                    <td>

                        <div class="action-buttons">

                            <button
                                class="action-btn edit"
                                title="Edit Volunteer"
                                onclick="editVolunteer(${volunteer.volunteer_id})"
                            >
                                ✏️
                            </button>


                            <button
                                class="action-btn delete"
                                title="Delete Volunteer"
                                onclick="deleteVolunteer(${volunteer.volunteer_id})"
                            >
                                🗑️
                            </button>

                        </div>

                    </td>

                </tr>
            `;
        }
    ).join("");
}


/* =========================
   ADD VOLUNTEER
========================= */

function openAddModal() {

    editingVolunteerId = null;

    document.getElementById("modal-title").textContent =
        "Add Volunteer";

    document.getElementById("modal-subtitle").textContent =
        "Register a new emergency volunteer";

    document.getElementById("save-button").textContent =
        "Save Volunteer";


    document.getElementById("volunteer-form").reset();

    document.getElementById("volunteer-id").value = "";

    document.getElementById("availability-status").value =
        "AVAILABLE";

    hideModalError();


    document
        .getElementById("volunteer-modal")
        .classList.add("show");
}


/* =========================
   EDIT VOLUNTEER
========================= */

function editVolunteer(volunteerId) {

    const volunteer = allVolunteers.find(
        item =>
            Number(item.volunteer_id) === Number(volunteerId)
    );


    if (!volunteer) {

        showError("Volunteer record not found.");

        return;
    }


    editingVolunteerId = volunteer.volunteer_id;


    document.getElementById("modal-title").textContent =
        "Edit Volunteer";

    document.getElementById("modal-subtitle").textContent =
        "Update volunteer information";

    document.getElementById("save-button").textContent =
        "Update Volunteer";


    document.getElementById("volunteer-id").value =
        volunteer.volunteer_id;

    document.getElementById("user-id").value =
        volunteer.user_id;

    document.getElementById("skills").value =
        volunteer.skills || "";

    document.getElementById("availability-status").value =
        volunteer.availability_status || "AVAILABLE";

    document.getElementById("emergency-contact").value =
        volunteer.emergency_contact || "";


    hideModalError();


    document
        .getElementById("volunteer-modal")
        .classList.add("show");
}


/* =========================
   SAVE / UPDATE
========================= */

async function saveVolunteer(event) {

    event.preventDefault();

    hideModalError();


    const userId =
        Number(
            document.getElementById("user-id").value
        );

    const skills =
        document.getElementById("skills").value.trim();

    const availabilityStatus =
        document.getElementById("availability-status").value;

    const emergencyContact =
        document
            .getElementById("emergency-contact")
            .value
            .trim();


    /* FRONTEND VALIDATION */

    if (!userId || userId < 1) {

        showModalError(
            "Please enter a valid User ID."
        );

        return;
    }


    const validStatuses = [
        "AVAILABLE",
        "ASSIGNED",
        "UNAVAILABLE"
    ];


    if (!validStatuses.includes(availabilityStatus)) {

        showModalError(
            "Please select a valid availability status."
        );

        return;
    }


    const payload = {
        user_id: userId,
        skills: skills || null,
        availability_status: availabilityStatus,
        emergency_contact: emergencyContact || null
    };


    const saveButton =
        document.getElementById("save-button");

    saveButton.disabled = true;
    saveButton.textContent = editingVolunteerId
        ? "Updating..."
        : "Saving...";


    try {

        let response;


        if (editingVolunteerId) {

            response = await fetch(
                `${API_BASE_URL}/api/volunteers/${editingVolunteerId}`,
                {
                    method: "PUT",
                    headers: {
                        "Content-Type": "application/json"
                    },
                    body: JSON.stringify(payload)
                }
            );

        } else {

            response = await fetch(
                `${API_BASE_URL}/api/volunteers/`,
                {
                    method: "POST",
                    headers: {
                        "Content-Type": "application/json"
                    },
                    body: JSON.stringify(payload)
                }
            );
        }


        const data = await response.json();


        if (!response.ok) {

            throw new Error(
                data.detail ||
                "Failed to save volunteer"
            );
        }


        alert(
            editingVolunteerId
                ? "Volunteer updated successfully!"
                : "Volunteer created successfully!"
        );


        closeModal();

        await loadVolunteers();


    } catch (error) {

        console.error(
            "Error saving volunteer:",
            error
        );

        showModalError(error.message);


    } finally {

        saveButton.disabled = false;

        saveButton.textContent =
            editingVolunteerId
                ? "Update Volunteer"
                : "Save Volunteer";
    }
}


/* =========================
   DELETE VOLUNTEER
========================= */

async function deleteVolunteer(volunteerId) {

    const volunteer = allVolunteers.find(
        item =>
            Number(item.volunteer_id) === Number(volunteerId)
    );


    if (!volunteer) {

        showError("Volunteer record not found.");

        return;
    }


    const name =
        volunteer.full_name || "this volunteer";


    const confirmed = confirm(
        `Are you sure you want to delete volunteer "${name}"?`
    );


    if (!confirmed) {
        return;
    }


    try {

        const response = await fetch(
            `${API_BASE_URL}/api/volunteers/${volunteerId}`,
            {
                method: "DELETE"
            }
        );


        const data = await response.json();


        if (!response.ok) {

            throw new Error(
                data.detail ||
                "Failed to delete volunteer"
            );
        }


        alert("Volunteer deleted successfully!");


        await loadVolunteers();


    } catch (error) {

        console.error(
            "Error deleting volunteer:",
            error
        );

        showError(error.message);
    }
}


/* =========================
   MODAL
========================= */

function closeModal() {

    document
        .getElementById("volunteer-modal")
        .classList.remove("show");

    hideModalError();

    editingVolunteerId = null;
}


/* Close when clicking outside */

document.addEventListener("click", event => {

    const modal =
        document.getElementById("volunteer-modal");

    if (
        event.target === modal &&
        modal.classList.contains("show")
    ) {
        closeModal();
    }

});


/* Close with Escape */

document.addEventListener("keydown", event => {

    if (event.key === "Escape") {

        const modal =
            document.getElementById("volunteer-modal");

        if (modal.classList.contains("show")) {
            closeModal();
        }

    }

});


/* =========================
   STATUS HELPERS
========================= */

function getStatusClass(status) {

    switch (status) {

        case "AVAILABLE":
            return "badge-available";

        case "ASSIGNED":
            return "badge-assigned";

        case "UNAVAILABLE":
            return "badge-unavailable";

        default:
            return "";
    }
}


/* =========================
   DATE FORMAT
========================= */

function formatDate(value) {

    if (!value) {
        return "—";
    }


    const date = new Date(value);


    if (Number.isNaN(date.getTime())) {
        return escapeHTML(value);
    }


    return date.toLocaleDateString(
        "en-IN",
        {
            day: "2-digit",
            month: "short",
            year: "numeric"
        }
    );
}


/* =========================
   ERROR HANDLING
========================= */

function showError(message) {

    const errorBox =
        document.getElementById("page-error");


    if (!errorBox) {
        return;
    }


    errorBox.innerHTML =
        escapeHTML(message);

    errorBox.classList.add("show");
}


function hideError() {

    const errorBox =
        document.getElementById("page-error");


    if (errorBox) {
        errorBox.classList.remove("show");
        errorBox.textContent = "";
    }
}


function showModalError(message) {

    const errorBox =
        document.getElementById("modal-error");


    errorBox.textContent =
        message;

    errorBox.classList.add("show");
}


function hideModalError() {

    const errorBox =
        document.getElementById("modal-error");


    errorBox.classList.remove("show");
    errorBox.textContent = "";
}

File: backend/app/routes/volunteers.py
from fastapi import APIRouter, Depends, HTTPException
from pydantic import BaseModel
from typing import Optional
from sqlalchemy.orm import Session
from sqlalchemy import text

from app.database import get_db

router = APIRouter(prefix="/api/volunteers", tags=["Volunteers"])


class VolunteerCreate(BaseModel):
    user_id: int
    skills: Optional[str] = None
    availability_status: str = "AVAILABLE"
    emergency_contact: Optional[str] = None


class VolunteerUpdate(BaseModel):
    user_id: int
    skills: Optional[str] = None
    availability_status: str
    emergency_contact: Optional[str] = None


@router.get("/")
def get_all_volunteers(db: Session = Depends(get_db)):
    query = text("""
        SELECT
            v.volunteer_id,
            v.user_id,
            u.full_name,
            u.email,
            u.phone,
            v.skills,
            v.availability_status,
            v.emergency_contact,
            v.joined_at
        FROM volunteers v
        JOIN users u ON v.user_id = u.user_id
        ORDER BY v.volunteer_id DESC
    """)

    result = db.execute(query).mappings().all()

    return {
        "count": len(result),
        "volunteers": [dict(row) for row in result]
    }


@router.get("/{volunteer_id}")
def get_volunteer(
    volunteer_id: int,
    db: Session = Depends(get_db)
):
    query = text("""
        SELECT
            v.volunteer_id,
            v.user_id,
            u.full_name,
            u.email,
            u.phone,
            v.skills,
            v.availability_status,
            v.emergency_contact,
            v.joined_at
        FROM volunteers v
        JOIN users u ON v.user_id = u.user_id
        WHERE v.volunteer_id = :volunteer_id
    """)

    result = db.execute(
        query,
        {"volunteer_id": volunteer_id}
    ).mappings().first()

    if not result:
        raise HTTPException(
            status_code=404,
            detail="Volunteer not found"
        )

    return dict(result)


@router.post("/")
def create_volunteer(
    volunteer: VolunteerCreate,
    db: Session = Depends(get_db)
):
    valid_statuses = [
        "AVAILABLE",
        "ASSIGNED",
        "UNAVAILABLE"
    ]

    if volunteer.availability_status not in valid_statuses:
        raise HTTPException(
            status_code=400,
            detail="Availability status must be AVAILABLE, ASSIGNED, or UNAVAILABLE"
        )

    user_check = db.execute(
        text("""
            SELECT user_id
            FROM users
            WHERE user_id = :user_id
        """),
        {"user_id": volunteer.user_id}
    ).first()

    if not user_check:
        raise HTTPException(
            status_code=404,
            detail="User not found"
        )

    existing_volunteer = db.execute(
        text("""
            SELECT volunteer_id
            FROM volunteers
            WHERE user_id = :user_id
        """),
        {"user_id": volunteer.user_id}
    ).first()

    if existing_volunteer:
        raise HTTPException(
            status_code=400,
            detail="This user is already registered as a volunteer"
        )

    query = text("""
        INSERT INTO volunteers (
            user_id,
            skills,
            availability_status,
            emergency_contact
        )
        VALUES (
            :user_id,
            :skills,
            :availability_status,
            :emergency_contact
        )
    """)

    result = db.execute(
        query,
        {
            "user_id": volunteer.user_id,
            "skills": volunteer.skills,
            "availability_status": volunteer.availability_status,
            "emergency_contact": volunteer.emergency_contact
        }
    )

    db.commit()

    return {
        "message": "Volunteer created successfully",
        "volunteer_id": result.lastrowid
    }


@router.put("/{volunteer_id}")
def update_volunteer(
    volunteer_id: int,
    volunteer: VolunteerUpdate,
    db: Session = Depends(get_db)
):
    volunteer_check = db.execute(
        text("""
            SELECT volunteer_id
            FROM volunteers
            WHERE volunteer_id = :volunteer_id
        """),
        {"volunteer_id": volunteer_id}
    ).first()

    if not volunteer_check:
        raise HTTPException(
            status_code=404,
            detail="Volunteer not found"
        )

    valid_statuses = [
        "AVAILABLE",
        "ASSIGNED",
        "UNAVAILABLE"
    ]

    if volunteer.availability_status not in valid_statuses:
        raise HTTPException(
            status_code=400,
            detail="Availability status must be AVAILABLE, ASSIGNED, or UNAVAILABLE"
        )

    user_check = db.execute(
        text("""
            SELECT user_id
            FROM users
            WHERE user_id = :user_id
        """),
        {"user_id": volunteer.user_id}
    ).first()

    if not user_check:
        raise HTTPException(
            status_code=404,
            detail="User not found"
        )

    duplicate_user = db.execute(
        text("""
            SELECT volunteer_id
            FROM volunteers
            WHERE user_id = :user_id
            AND volunteer_id != :volunteer_id
        """),
        {
            "user_id": volunteer.user_id,
            "volunteer_id": volunteer_id
        }
    ).first()

    if duplicate_user:
        raise HTTPException(
            status_code=400,
            detail="This user is already registered as another volunteer"
        )

    query = text("""
        UPDATE volunteers
        SET
            user_id = :user_id,
            skills = :skills,
            availability_status = :availability_status,
            emergency_contact = :emergency_contact
        WHERE volunteer_id = :volunteer_id
    """)

    db.execute(
        query,
        {
            "volunteer_id": volunteer_id,
            "user_id": volunteer.user_id,
            "skills": volunteer.skills,
            "availability_status": volunteer.availability_status,
            "emergency_contact": volunteer.emergency_contact
        }
    )

    db.commit()

    return {
        "message": "Volunteer updated successfully",
        "volunteer_id": volunteer_id
    }


@router.delete("/{volunteer_id}")
def delete_volunteer(
    volunteer_id: int,
    db: Session = Depends(get_db)
):
    volunteer_check = db.execute(
        text("""
            SELECT volunteer_id
            FROM volunteers
            WHERE volunteer_id = :volunteer_id
        """),
        {"volunteer_id": volunteer_id}
    ).first()

    if not volunteer_check:
        raise HTTPException(
            status_code=404,
            detail="Volunteer not found"
        )

    try:
        db.execute(
            text("""
                DELETE FROM volunteers
                WHERE volunteer_id = :volunteer_id
            """),
            {"volunteer_id": volunteer_id}
        )

        db.commit()

    except Exception:
        db.rollback()

        raise HTTPException(
            status_code=400,
            detail="Volunteer cannot be deleted because the volunteer is being used by another record"
        )

    return {
        "message": "Volunteer deleted successfully",
        "volunteer_id": volunteer_id
    }